feat: Add ability to customize function dockerfile

chxlloyd opened the pull request:
Allows client to customize the function dockerfile and the build context that is applied for it.

bestickley commented on src/nextjs-build/nextjs-build.ts:
```diff
@@ -205,8 +208,20 @@ export class NextjsBuild extends Construct {
+
+    const filePathsToCopy = [
+      srcDockerfilePath,
+      join(__dirname, "add-cache-handler.mjs"),
+      join(__dirname, "cache-handler.cjs"),
```
do we need to `join(__dirname, ...` here if we're already calling `basename` below?

chxlloyd replied:
This is for the full filepath of cdk-nextjs to copy into the buildContext so it has access to add-cache-handler and cache-handler.  Without the join, it would copy from wherever this is being invoked.

bestickley replied:
oh duh, I missed line 220 below. thankyou.

bestickley commented on src/nextjs-build/nextjs-build.ts:
```diff
@@ -291,8 +308,16 @@ export class NextjsBuild extends Construct {
+      join(__dirname, "..", "..", "lib", "nextjs-build");
     const dockerImageAsset = new DockerImageAsset(this, "Image", {
+      directory: buildContext,
+      exclude: ["*", `!${dockerfileName}`, "!add-cache-handler.mjs"],
```
do you think we should change default behavior of exclude? by default we ignore everything and allow list the dockerfilename and add-cache-handler.mjs but I could foresee this frustrating end users when they're trying to `ADD` a file from their build context and it's being ignored by default. what do you think?

same applies for below.

chxlloyd replied:
Based on my understanding, yeah a default exclude all seems like it would be cumbersome.  However, with these updates they can override them if need be.

bestickley commented on src/nextjs-build/nextjs-build.ts:
```diff
@@ -327,18 +355,16 @@ export class NextjsBuild extends Construct {
         `!${dockerfileName}`,
         "!add-cache-handler.mjs",
         "!cache-handler.cjs",
+        ...(this.props.overrides?.nextjsFunctionsAssetImageCodeProps?.exclude ??
```
let's make this exclude consistent with others.

chxlloyd replied:
Updated.

bestickley commented on src/nextjs-build/nextjs-build.ts:
```diff
@@ -350,27 +376,31 @@ export class NextjsBuild extends Construct {
       file: dockerfileName,
       ignoreMode: IgnoreMode.DOCKER,
       ...this.props.overrides?.nextjsAssetDeploymentAssetImageCodeProps,
+      buildArgs: {
```
is exclude missing from here?

chxlloyd replied:
Updated.

bestickley commented on src/nextjs-build/nextjs-build.ts:
```diff
@@ -66,6 +66,9 @@ export interface NextjsBuildOverrides {
   readonly nextjsContainersDockerImageAssetProps?: OptionalDockerImageAssetProps;
   readonly nextjsFunctionsAssetImageCodeProps?: AssetImageCodeProps;
   readonly nextjsAssetDeploymentAssetImageCodeProps?: AssetImageCodeProps;
+  readonly functionsImageBuildContext?: string;
+  readonly assetsDeploymentImageBuildContext?: string;
+  readonly containersImageBuildContext?: string;
```
Let's add TS Doc to each adjusting description as necessary.:
```
/**
* Absolute path to [build context](https://docs.docker.com/build/concepts/context/) which should be 
* directory containing Dockerfile which will define the container image used to run the Next.js 
* {function|container}
* @example
* join(import.meta.dirname, "..", "my-build-context")
*/
```

chxlloyd replied:
Updated.

cdklabs-automation committed:
chore(deps): upgrade dependencies (#107)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdklabs/cdk-nextjs/actions/runs/14961110273

------

*Automatically created by projen via the "upgrade-main" workflow*

Signed-off-by: github-actions <github-actions@github.com>
Co-authored-by: github-actions <github-actions@github.com>

bestickley committed:
ci: e2e tests (#110)

* ci: e2e tests

* chore: update jsii

* fix: github action name convention

* fix: setup pnpm

* fix: pnpm version

* fix: yaml multline

* fix: build cdk-nextjs first

* fix: e2e-test.yaml

* fix: actions checkout

* fix: gh composite action secret and shell

* fix: remove shell

* fix: secret use in composite action

* fix: cwd

* fix: cfn stacks must start with character

* fix: remove destory on pr

* fix: test:ci

* fix: StackProps env

* fix: outputs.json

* centralize run-e2e-tests-on-example action

* fix: add shell

* fix: E2E_BASE_URL

* fix: properly escape jq command

* fix: jq command

* refactor: remove duplicative base url export

* fix: cdk-nextjs=1 cookie for regional containers

* fix: http for regional containers

bestickley committed:
ci: fix e2e test clean up (#114)

* ci: build cdk-nextjs for cleanup

* fix: trigger e2e-test workflow after labeling pr

* fix: permissions

* fix: skip-e2e-tests label

* fix: debug s3 key

gbone-restore committed:
fix:  basePath not being applied to static assets in S3 and Cloudfront behaviors (#102)

* Fix for basePath not being applied to static assets in S3 and Cloudfront behaviors

* fix: api docs

---------

Co-authored-by: gbone24 <4950487+gbone24@users.noreply.github.com>
Co-authored-by: Ben Stickley <stickb@amazon.com>

bestickley committed:
ci: parallel cleanup (#116)

* fix: parallel cleanup

* ci: continue on error

bestickley committed:
ci: fix account id (#117)

bestickley committed:
ci: e2e tests on fork pr (#118)

* ci: e2e tests on fork pr

* ci: check-fork and wait for approval

* ci: simplify check-fork

* ci: remove need for auto-approve environment

* ci: fix needs wait-for-approval

Commit:
feat: ability to customize compute image build context and Dockerfile

Commit:
fix: custom cacheHandler

Commit:
feat!: replace NextjsInvalidation with NextjsPostDeploy, add efs and s3 pruning

Commit:
fix!: remove NextjsRevalidation

Commit:
feat: reduce builder image size

Commit:
fix!: deprecate relativePathToWorkspace, replace with relativePathToPackage

Commit:
docs: turborepo example

Commit:
docs: breaking changes, update examples, e2e tests

bestickley committed:
Merge branch 'main' into feature/custom-docker-filepath

Commit:
style: import order

Commit:
docs: comment updates
